### ISSUE TITLE
Ranges of fixes

### ISSUE THREAD

**Issue by romainmenke** [stylelint/stylelint#7788]
See: 
- https://github.com/stylelint/stylelint/issues/7784
- https://eslint.org/docs/latest/integrate/nodejs-api#-lintmessage-type

------

## Warnings and fixes have different target audiences

**A warning is a message to a human** and it is sufficient to give a general range and a helpful message. Ideally we always have perfect ranges, but because we know that there will always be a person to interpret the message we can have less than ideal reports.

Example : https://stylelint.io/demo/#N4Igxg9gJgpiBcIB0AzCEAEwA6A7DGADgIZRQCWuA5gLQA2MKALvBgIwAMhAHgNx4FIdCACdWImFH74ipCtRojyVABYt2XPnmxMARsTABrKiIgBXXFBpDRrEzBi5pBEmUq0mEQq049nstwVdCCZPAFsfTWkAXxAAGhAUcgYAOWIwuEQYbnTCBiQwAGdC+PAIXCSqBBAcGWwQbKZHKEL61gBtAQIMesKmAE8GOkoma3LKmj7iS2IRKHqugF08WITICuUAMVEw4iZqgCtC8tLYQhLEWoJegaGRtp6QOj2YPvq4rpvBmGHcUfWJlMZnMHvVnk03iAViBokA

<img width="411" alt="Screenshot of the playground example linked above" src="https://github.com/stylelint/stylelint/assets/11521496/536cd0cf-b660-493b-9988-17d479bda617">

Only the property name of the last declaration is part of the warning range.

**A fix will be a message to another tool** and it must be very accurate for the fix to be applied correctly.

## Non-consecutive fix ranges

We can handle non-consecutive fix ranges in two ways.
Either we extend the range or we have multiple ranges.

### Multiple ranges

This is more granular and makes it possible to have non-consecutive ranges that do not conflict with fixes on the interleaved source.

e.g. Another fixer that changes `green` to `limegreen` in the example above.

However it implies significant complexity because only the first entry in the array of ranges would be accurate. All subsequent entries would become incorrect after applying the first splice.

### Extended range

This is less accurate and requires that fix ranges encompass the entire affected area, included unchanged interleaved bits of source code.

This is much simpler.

In the example given above it might be fine to extend the range of the fix to that of the whole parent. This might be much more than is really needed, but it can never be wrong.

### My preference at this time

I currently favor having an extended range
- (+) it is simpler for us and for tool authors
- (+) a single range per fix is less prone to bugs, both for us and for tool authors
- (+) it makes it possible to align our public API with that of ESLint
- (-) it is more convoluted, less obvious for rule authors (you need to know that range and replacement text should encompass the entire area)

## `index/endIndex` vs. `start/end` with `line,column`

A separate but equally important aspect is `index/endIndex` vs. `start/end` with `line,column`.
I don't have a good understanding of what the benefits are of either.
I know that some tools expect one or the other and that you can calculate one from the other.

ESLint uses `index/endIndex`.

I have no preference or strong opinion about this.

**Comment by Mouvedia** [comment]
> Warnings and fixes have different target audiences

During the process of the past refactors I did change the warning range _once_.
It was motivated by the fact that the previously reported warning range could be confusing in certain cases.
see https://github.com/stylelint/stylelint/pull/7758#issue-2351951358
This could be reverted but I believe it's better that way.

> multiple ranges
> extended ranges

I am fine with either choice.
Supporting multiple ranges is just a matter of putting it in the type.
It will become a real bother once we will have to test those ranges but for now it's the most future-proof choice of the 2 IMHO.

> index/endIndex vs. start/end with line,column

I think this is irrelevant to the matter unless you are advocating for not having a fallback.
I support having a fallback range for all the rules that already have a range which would match the new _old text_ range that we are introducing.

**Comment by ybiquitous** [comment]
@romainmenke Thanks for opening the discussion. To make sure, this issue discusses how a new property in `Warning` should be constructed for autofix, right?

https://github.com/stylelint/stylelint/blob/e95d114e2c57ef0426ab7637d4a9e123a203fce4/types/stylelint/index.d.ts#L493-L515

**Comment by romainmenke** [comment]
In a way, yes :)
But more specifically what the range values are for such a new property.

This is not about the existing range values for warnings.

I think that these are conceptually different things, even if they align in practice for most rules.

-----

I want to explore the unique issues and constraints that exist for fix ranges and formulate guidance for future refactors.

Doing this exploratory work through refactors and code review is expensive (time and effort)

**Comment by Mouvedia** [comment]
@ybiquitous 

> [Compiling the fixable rules that currently don't return a warning range that would match the old text range in every cases is our next task.](https://github.com/stylelint/stylelint/issues/7192#issuecomment-2184053753)

**Comment by ybiquitous** [comment]
@romainmenke @Mouvedia Thanks for your responses. I understand this issue is derived from #7192. And, it focuses on a new property for fix ranges that will be added to the `Warning` type.

For example, in this issue, should we discuss something like this?

```ts
type FixRange = [number, number]; // [startIndex, endIndex]
// or
type FixRange = { // same as `Range`
	start: { line: number, column: number };
	end: { line: number, column: number };
};

type Warning = {
	fix?: {
		replacement: string;
		ranges: Array<FixRange>;
	};
};
```

Ref:
- https://github.com/stylelint/stylelint/issues/7192#issuecomment-1726698251
- https://github.com/stylelint/stylelint/issues/7192#issuecomment-2163732191
- https://eslint.org/docs/latest/integrate/nodejs-api#-editinfo-type

**Comment by Mouvedia** [comment]
I am of the opinion that we should keep "multiple ranges" in the type.
i.e. `| Range[]`
But we should support "extended ranges" for every rule first.
i.e. once all rules have been reviewed we can remove `| Range[]` if we feel like it's not worth the effort for the few rules that it would benefit or migrate the rules in question
I would find it weird that we don't dogfood a new feature.
i.e. if it's not good enough for a core rule then it shouldn't be pushed to the plugin authors

The idea of "multiple ranges" for the same fix has been introduced by Romain. He has sound arguments against it now.
I don't mind either way. I am just saying that it will harder to add something to the type after-the-fact.

**Comment by ybiquitous** [comment]
Sorry, I may not understand **"Multiple ranges"** and **"Extended range"** well. 😓 

Can anyone provide a simple TypeScript code to understand both easily?

**Comment by Mouvedia** [comment]
I am referring to the terms used by Romain in [OP](https://github.com/stylelint/stylelint/issues/7788#issue-2367918084).

**Comment by romainmenke** [comment]
Single range:

```ts
type Range = {
	start: { line: number, column: number };
	end: { line: number, column: number };
};

type EditInfo = { // same as `Range`
	replacement: string;
	range: Range;
};

type Warning = {
	fix?: EditInfo;
};
```

Multiple ranges:

```ts
type Range = {
	start: { line: number, column: number };
	end: { line: number, column: number };
};

type EditInfo = { // same as `Range`
	replacement: string;
	range: Range;
};

type Warning = {
	fix?: Array<EditInfo>;
};
```

_Notice how each `range` has exactly one `replacement` in both options._

------

https://stylelint.io/demo/#N4Igxg9gJgpiBcIB0AzCEAEwA6A7DGADgIZRQCWuA5gLSUA2lMNAzgC7EBOb8GAjAAZCADwDceApHoROvTjCjj8RUhWp1cjXMxi4ovQSKUBfEABoQKcvRgA5YgFs4iGMMeEbSMCxbnwEXCsqBBAcZWwQVzZdKBYI3gBtCQIMCPYATxstNhpIQPJadmI9LigI5IBdPFMLPKCAMRkHYjYQgCsWAL9YQl9EMII0tkyYbPjUkHoWmHYIs2ShkezcgKDWDhLOMoQJqejZkGqQYyA

```css
.foo {
  padding-inline-start: 10px;
  color: red;
  padding-inline-end: 10px;
}
```

To describe the fix for the above CSS you only have two options:
- list multiple replacements (multiple ranges)
- describe a larger area to be replacement ("extend" the range)

Multiple ranges: 

```ts
{
	fix?: [
		{
			replacement: "padding-inline: 10px 10px;"
			range: {
				start: { line: 2, column: 3 },
				end: { line: 2, column: 30 },
			}
		},
		{
			replacement: ""
			range: { // Describes the position as if the previous fix was not applied. Shifts of the ranges will be needed if applying edits in sequence.
				start: { line: 4, column: 0 }, // including the indent
				end: { line: 4, column: 29 }, // including the newline
			}
		},
	]
};
```

Extended range: 

```ts
{
	fix?: {
		replacement: "padding-inline: 10px 10px;\n  color: red;"
		range: {
			start: { line: 2, column: 3 },
			end: { line: 4, column: 28 },
		}
	}
};
```

**Comment by romainmenke** [comment]
Side note:

This issue also exists twice and we don't need to pick the same option twice.
- the `fix` property on `Warning` will need to expose range info
- the `fix()` callback in `report` will need to return range info

We could allow rule authors to return an array in `fix()` while at the same time only exposing a single range in `Warning.fix`

**Comment by ybiquitous** [comment]
@romainmenke Thank you so much. I now understand what you mean. 👍🏼 

At this moment, I feel "Extended range" is better because of a single `EditInfo`.

@Mouvedia What do you think?

**Comment by Mouvedia** [comment]
I am fine with both options. If we change our minds, it will probably only be supported by either core rules and new plugin rules.
i.e. the upcoming wave of refactors by plugin authors won't have that option

> the fix() callback in report will need to return range info

@ybiquitous can you chime in on https://github.com/stylelint/stylelint/issues/7784#issuecomment-2184094675 and give your preference?
i.e. 1, 2 or 3 (but do not take into account the array matter discussed here)

**Comment by ybiquitous** [comment]
To make sure, we have eventually agreed on the following type:

```ts
type EditInfo = {
	replacement: string;
	range: Range;
};

type Warning = {
	fix?: EditInfo;
};
```

Based on this decision, I'll consider the `report()` API discussed in #7784.

**Comment by romainmenke** [comment]
Yes, at this time I think that is the best option.

- each `Warning` has zero or one `fix`
- each `fix` has exactly one `replacement`
- each `fix` has exactly one `range`

Closing this issue as we seem to have consensus :)
We can re-open if we learn new things that require revisiting the proposed API in https://github.com/stylelint/stylelint/issues/7788#issuecomment-2186850056

**Comment by romainmenke** [comment]
Using larger ranges that encompass the entire affected region might be unintuitive for users in some cases.

The ranges are relevant to configuration commands in comments.

I think the most surprising case is when a range starts right before a newline, as newlines don't get the squiggly lines in editors.

```
.a,
   ^ report range starts here, at the newline
.b {
^^ squiggly line is only visible here
  color: red;
}
``` 

I don't think this is a blocking issue, so not re-opening at this time, but good to keep in mind.

**Comment by romainmenke** [comment]
I update a plugin we use to sort declarations to utilize the new `fix()` function and this required having a single range that spans multiple declarations.

You can see the source code here: https://github.com/mrhenry/stylelint-mrhenry/blob/6939dd319ec125ae7ae9b3939bf106e9f39b452a/packages/prop-order/index.mjs#L106-L129

What I did:
- use the parent node as the `node` param in `report()`
- take the start source position of the first declaration affected by the error
- take the end source position of the last declaration affected by the error
- offset both source positions with the start source position of the parent
- use the result as `index` and `endIndex`

This was fairly straightforward and seems to give correct results.
